Added support for Qanary component logging

MindMaster98 opened the pull request:
Providing a new package "logging" with an abstract class "QanaryComponentLogger", which can be used to implement custom logging mechanisms. By default, the "MLFlowLogger" class is provided allowing Qanary components to log their results to a MLFlow instance.

## Test
To test the logging run the `test_logging.py` script. Therefore, a Docker Image with a configured MLFlow server is provided in the directory `tests/logging_test_mlflow_instance`. Before starting the Docker Container it is required to set the following environment variables for the Docker-Compose file *and* the host system.

```bash
MLFLOW_URI=http://localhost:5000
MLFLOW_HOST=localhost
MLFLOW_PORT=5000
MLFLOW_PORT_ARTIFACT=5001
MLFLOW_USERNAME=mlflow
MLFLOW_PASSWORD=mlflow
```

Perevalov commented:
1. My Python environment doesn't accept this kind of type specifications: `list[dict[str, str | float]]`. Should I install anything in addition?

Perevalov commented:
2. Could you update README.md with instruction on how to call the logger within a component?

MindMaster98 commented:
@Perevalov I updated the README with an example how to use the MLFlowLogger. `list[dict[str, str | float]]` is only supported by Python 3.9+, so i replaced it with regular typing.

Perevalov commented:
3. Please, also add `MLFLOW_URI` parameter and its value to your description example

MindMaster98 commented:
@Perevalov I updated the test for the logging so that no additional MLFlow servers needs to be started to run the tests.

Perevalov commented:
@MindMaster98 I tried to run the tests and got the following:

```log
alex@ins-alex-ThinkPad-T15g-Gen-1:~/playground/qanary_helpers/tests$ python3 test_logging.py 
/usr/lib/python3/dist-packages/paramiko/transport.py:219: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
EEE
======================================================================
ERROR: test_annotation_logging (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_logging.py", line 86, in test_annotation_logging
    run_id = self.logger.log_annotation('MusicianAnnotator:latest', 'How much is the fish?', 'Scooter',
TypeError: log_annotation() missing 1 required positional argument: 'qanary_graph_id'

======================================================================
ERROR: test_test_logging (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_logging.py", line 58, in test_test_logging
    run_ids = self.logger.log_test_results([
  File "/home/ins-alex/.local/lib/python3.8/site-packages/qanary_helpers/logging/component_logging.py", line 124, in log_test_results
    mlflow.log_param(test_param, question[test_param])
KeyError: 'docker_image_tag'

======================================================================
ERROR: test_train_logging (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_logging.py", line 32, in test_train_logging
    run_id = self.logger.log_train_results('test:latest', train, test, {'C': 0.5, 'balance': 'weighted'},
TypeError: log_train_results() takes 10 positional arguments but 12 were given

----------------------------------------------------------------------
Ran 3 tests in 0.184s

FAILED (errors=3)
```

I saw that the code is correct, but something does not work. Did you try it with Python 3.8 also?

Perevalov commented:
@MindMaster98 Thanks, the local tests ran correctly, please add `mlruns/` to the `.gitignore` and then I merge.

MindMaster98 commented:
@Perevalov I updated the `.gitignore`. The logging can be merged now.

Perevalov reviewed (commented):
Looks good
